File: Brazil.java
import java.io.IOException;
import java.io.RandomAccessFile;

public class Brazil {
    private static void show(final long[][] flag, final String chars, final String slogan) {
        int c_ = 0;
        int li = 0;
        for (int i = 0; i < 15; i++) {
            for (int j = 0; j < 41; j++) {
                int c = 0;
                final int li_ = li;
                while (li < 4) {
                    c = chars.charAt(li);
                    if ((flag[li][i] >> j & 1) == 0) {
                        break;
                    }
                    ++li;
                }
                li = li_;
                if (c == chars.charAt(chars.length() - 1)) {
                    c = slogan.charAt(c_++);
                }
                System.out.print(Character.toString(c));
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        final long[][] flag = new long[4][15];
        try (final RandomAccessFile data = new RandomAccessFile("flag", "r")) {
            final byte[] bytes = new byte[6];
            for (int i = 0; i < 4; i++) {
                for (int j = 0; j < 15; j++) {
                    data.readFully(bytes);
                    flag[i][j] = 0L;
                    for (int bi = 0; bi < 6; bi++) {
                        flag[i][j] |= (long) (bytes[5 - bi] & 0xFF) << (8 * bi);
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        show(flag, ".#` ", " ORDEM E PROGRESSO ");
    }
}

File: Brazil.kt
import java.io.RandomAccessFile

private fun show(flag: Array<LongArray>, chars: String, slogan: String) {
    var c_ = 0
    var li = 0
    repeat(15) { i ->
        repeat(41) { j ->
            var c = chars[li]
            val li_ = li
            while (li < 4) {
                c = chars[li]
                if (flag[li][i].shr(j).and(1) == 0L) {
                    break
                }
                ++li
            }
            li = li_
            if (c == chars.last()) {
                c = slogan[c_++]
            }
            print(c)
        }
        println()
    }
}

fun main() {
    val flag = Array(4) { LongArray(15) }
    RandomAccessFile("flag", "r").use { data ->
        val bytes = ByteArray(6)
        repeat(4) { i ->
            repeat(15) { j ->
                data.readFully(bytes)
                flag[i][j] = 0L
                repeat(6) { bi ->
                    val mask = ((bytes[5 - bi].toLong().and(0xFF))).shl(8 * bi)
                    flag[i][j] = flag[i][j].or(mask)
                }
            }
        }
    }
    show(flag, ".#` ", " ORDEM E PROGRESSO ")
}
